### ISSUE TITLE
Cannot connect to remote node from Python, after reboot

### ISSUE THREAD

**Issue by vegabook** [esl/Pyrlang#20]
After rebooting a Python-side node, it will not connect to a remote node on the first attempt. One must first connect to an erlang/elixir node on the same physical machine, then after disconnecting, one can then connect to remote nodes. 

Remote node running Erlang/Elixir: 192.168.1.199
Local node running Pyrlang: 192.168.1.192

code:

```
import gevent
from gevent import monkey
monkey.patch_all()
import pdb

import Pyrlang
from Pyrlang import Atom

def main():
    node = Pyrlang.Node("py@192.168.1.192", "COOKIE")
    node.start()

    # Attempt to send something will initiate a connection before sending
    pid = node.register_new_process(None)
    # To be able to send to Erlang shell by name first give it a registered
    # name: `erlang:register(shell, self()).`
    node.send(pid, (Atom('erl@192.168.1.199'), Atom('shell')), Atom('hello'))

    while True:
        # Sleep gives other greenlets time to run
        gevent.sleep(0.1)

if __name__ == "__main__":
    main()
```

Start Elixir on remote node:

`pi@flotilla:~/www/mspiex --name erl@192.168.1.199 --cookie COOKIE`

Run code on local node:

```
tbrowne@cubietruck ~/c/e/p/p/python> python3 test6.py
Dist: Listening for dist connections on port 36539
EPMD: Connecting 127.0.0.1:4369
EPMD: connection error: [Errno 111] Connection refused
EPMD: Connecting 127.0.0.1:4369
EPMD: connection error: [Errno 111] Connection refused
```

Now change code so remote node is same computer so both 192.168.1.192:

```
tbrowne@cubietruck ~/c/e/p/p/python> cat test6.py
import gevent
from gevent import monkey
monkey.patch_all()
import pdb

import Pyrlang
from Pyrlang import Atom

def main():
    node = Pyrlang.Node("py@192.168.1.192", "COOKIE")
    node.start()

    # Attempt to send something will initiate a connection before sending
    pid = node.register_new_process(None)
    # To be able to send to Erlang shell by name first give it a registered
    # name: `erlang:register(shell, self()).`
    node.send(pid, (Atom('erl@192.168.1.192'), Atom('shell')), Atom('hello'))

    while True:
        # Sleep gives other greenlets time to run
        gevent.sleep(0.1)

if __name__ == "__main__":
    main()
```

Start local machine Elixir node:
```

tbrowne@cubietruck ~> iex --name erl@192.168.1.192 --cookie COOKIE
Erlang/OTP 20 [erts-9.1] [source] [smp:2:2] [ds:2:2:10] [async-threads:10] [hipe] [kernel-poll:false]

Interactive Elixir (1.5.3) - press Ctrl+C to exit (type h() ENTER for help)
iex(erl@192.168.1.192)1> 
```

Start Pyrlang on local machine:

```
tbrowne@cubietruck ~/c/e/p/p/python> python3 test6.py
Dist: Listening for dist connections on port 41409
EPMD: Connecting 127.0.0.1:4369
EPMD: Socket connected
EPMD: sending ALIVE2 req ('py@192.168.1.192', 77, (5, 5))
EPMD: Connected successfully (creation 1)
send -> (atom'erl@192.168.1.192', atom'shell'): hello
Node._send_remote shell <- hello
Node: connect to node erl@192.168.1.192
Dist-out: Send_name b'n\x05\x05\x00\x03\x0f\xbcpy@192.168.1.192'
Connection to ('192.168.1.192', 34369) established
Node: wait for 'node_connected'
Out-connection established with erl@192.168.1.192
Node: connected
```

Connected. Ping and RPC all work as expected. Now try again with first version of code, ie back to 192.168.1.199
```

tbrowne@cubietruck ~/c/e/p/p/python> python3 test6.py
Dist: Listening for dist connections on port 40479
EPMD: Connecting 127.0.0.1:4369
EPMD: Socket connected
EPMD: sending ALIVE2 req ('py@192.168.1.192', 77, (5, 5))
EPMD: Connected successfully (creation 2)
send -> (atom'erl@192.168.1.199', atom'shell'): hello
Node._send_remote shell <- hello
Node: connect to node erl@192.168.1.199
Dist-out: Send_name b'n\x05\x05\x00\x03\x0f\xbcpy@192.168.1.192'
Connection to ('192.168.1.199', 41883) established
Node: wait for 'node_connected'
Out-connection established with erl@192.168.1.199
Node: connected
```

Connected. Ping and RPC work as expected. I have tried this on Linux, ARM instructions set and Intel instruction set, and even on Linux to Windows. Always behaves the same way. The python node will never connect to an Elixir or Erlang node on another IP address, unless you have first, once, connected to the same IP address as the Pyrlang node is running on. Thereafter you can stop all processes and start again to remote node and it works. Something weird is happening on ports initialization or something and I don't know what. 

Basically, on first attempt after any reboot of the Pyrlang side remote-address node, you cannot connect to other-address Elixir/Erlang. You first have to connect to same-address elixir/erlang, and then for rest of life of session remote IP connections will work.

**Comment by kvakvs** [comment]
Pyrlang is not able to run EPMD. It probably should be able to find it and run it. Your symptoms (EPMD connection refused) mean EPMD is not running.

Other solution would be to provide an own EPMD replacement, it is a very simple C program used by Erlang/Elixir as a running node registry and is started by Erlang/Elixir automatically.

Here is documentation and starting http://erlang.org/doc/man/epmd.html you can basically do `epmd -d` and it will be sufficient for your needs, even if its already running the new one should exit because EPMD port is in use.
